instrumentation/opentelemetry_bandit/lib/opentelemetry_bandit.ex

```diff
@@ -1,178 +1,515 @@
+  SemConv attribute.
+
+  Example:
+  ```
+  opt_ins = [{SemConv.HTTPAttributes.http_request_body_size(), true}]`
+  OpentelemetryBandit.setup(opt_in_attrs: opt_ins)
+  ```
+
+  Request and response header attributes are opt-in and can be set with the
+  `request_headers` and `response_headers` options. Values should be lower-case.
+
+  ### Public Endpoint
+
+  By default, the endpoint is handled as non-public. Setting an endpoint as
+  public will result in any propagated trace to be added as a link, not a
+  continuation of an existing trace. The `public_endpoint` option should be set
+  to `true` if an endpoint only accepts public traffic to prevent missing root spans.
+
+  In a mixed traffic environment, an MFA can be supplied to determine whether to
+  treat a request as public. This function is executed on every request so refrain
+  from expensive operations such as lookups to external systems. The function must
+  be a predicate function of arity-2, accepting the `conn` from the request as
+  the first argument and the second being any user-supplied options. Any dynamic
+  information used in comparisons should be supplied at setup time for efficiency.
+
+  Example:
+  ```
+  defmodule PublicEndpoint do
+    def is_public_request?(conn, opts) do
+      # return true if request was public
+    end
+  end
+
+  OpentelemetryBandit.setup(public_endpoint_fn: {PublicEndpoint, :is_public_request?/2, []})
+  ```
+  """
+
   @doc """
   Initializes and configures the telemetry handlers.
+
+  Supported options:\n#{NimbleOptions.docs(@options_schema)}
   """
   @spec setup(any) :: :ok
-  def setup(_opts \\ []) do
-    :telemetry.attach(
-      {__MODULE__, :request_stop},
-      [:bandit, :request, :stop],
-      &__MODULE__.handle_request_stop/4,
-      %{}
+  def setup(opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @options_schema)
+
+    config =
+      opts
+      |> Enum.into(%{})
+      |> Map.update!(:opt_in_attrs, fn attrs ->
+        attrs
+        |> Enum.filter(&elem(&1, 1))
+        |> Keyword.keys()
+      end)
+      |> Map.update!(:scheme_headers, &Enum.reverse/1)
+      |> Map.update!(:client_address_headers, &Enum.reverse/1)
+
+    :telemetry.attach_many(
+      {__MODULE__, config.handler_id},
+      [
+        [:bandit, :request, :start],
+        [:bandit, :request, :stop],
+        [:bandit, :request, :exception]
+      ],
+      &__MODULE__.handle_request/4,
+      config
     )
+  end
+
+  @doc false
+  def handle_request([:bandit, :request, :start] = event, measurements, meta, config) do
+    handle_request_start(event, measurements, meta, config)
+  end
+
+  @doc false
+  def handle_request([:bandit, :request, :stop] = event, measurements, meta, config) do
+    handle_request_stop(event, measurements, meta, config)
+  end
+
+  @doc false
+  def handle_request([:bandit, :request, :exception] = event, measurements, meta, config) do
+    handle_request_exception(event, measurements, meta, config)
+  end
+
```

Comment:
You can't `attach_many` with different function names. I don't mind doing the parameter-manipulation as you suggested. I just don't think pattern matching the event on functions that combined span hundreds of lines helps with readability or code navigation. It also results in even longer function heads since you have to match upwards of 30 chars